### PR TITLE
[DOCS] Deploy Kyuubi engines on Kubernetes

### PR DESCRIPTION
**Pull request by zwangsheng**
### _Why are the changes needed?_

Add content to kyuubi_on_kubernetes.md about deploying kyuubi on k8s

### _How was this patch tested?_
- [ ] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [X] [Run test](https://kyuubi.readthedocs.io/en/latest/tools/testing.html#running-tests) locally before make a pull request

### CODE REVIEW

**Comment by zwangsheng** [comment]
cc @yaooqinn  @ulysses-you @pan3793

File: docs/deployment/on_kubernetes.md [hunk 1]
```diff
@@ -5,3 +5,111 @@
...
+`k8s://https://<k8s-apiserver-host>:<k8s-apiserver-port>`
+
+You can use cmd `kubectl cluster-info` to get api-server host and port.
+
+### Spark Image
+
+Kubernetes requires users to supply images that can be deployed into containers within pods.
+
+Spark also ships within a bin/docker-image-tool.sh script that can be used to build and publish the Docker images to use
+with the K8S backend.
```
Comment (pan3793) [inline docs/deployment/on_kubernetes.md:50]: `K8s`, K should be uppercase, s should be lowercase

File: docs/deployment/on_kubernetes.md [hunk 2]
```diff
@@ -5,3 +5,111 @@
...
+watch executor pods.
+
+In both cases, you need to figure out whether you have the permissions under the corresponding namespace. You can use
+following cmd to create serviceAccount (You need to have the kubeconfig which have the create serviceAccount permission)
+.
+
+```shell
+#create serviceAccount
+kubectl create serviceaccount spark -n <namespace>
+#binding role
```
Comment (pan3793) [inline docs/deployment/on_kubernetes.md:82]: nit: one space after #

File: docs/deployment/on_kubernetes.md [hunk 3]
```diff
@@ -5,3 +5,111 @@
...
+* Read about [Running Spark On Kubernetes](http://spark.apache.org/docs/latest/running-on-kubernetes.html)
+* An active Kubernetes Cluster
+* KubeCtl
+* KubeConfig of the target cluster
+
+## Configurations
+
+### Test Cluster
+
+You can use the blew shell code to test your cluster whether it is normal or not.
```
Comment (pan3793) [inline docs/deployment/on_kubernetes.md:22]: typo: blew

File: docs/deployment/on_kubernetes.md [hunk 4]
```diff
@@ -5,3 +5,111 @@
...
+
+`k8s://https://<k8s-apiserver-host>:<k8s-apiserver-port>`
+
+You can use cmd `kubectl cluster-info` to get api-server host and port.
+
+### Spark Image
+
+Kubernetes requires users to supply images that can be deployed into containers within pods.
+
+Spark also ships within a bin/docker-image-tool.sh script that can be used to build and publish the Docker images to use
```
Comment (pan3793) [inline docs/deployment/on_kubernetes.md:49]: `bin/docker-image-tool.sh` should be quoted with `

File: docs/deployment/on_kubernetes.md [hunk 5]
```diff
@@ -5,3 +5,111 @@
...
 
 # Running Kyuubi on Kubernetes
+
+## Requirements
+
+When you want to run Kyuubi's Spark SQL engines on Kubernetes, you'd better have cognition upon the following things.
+
+* Read about [Running Spark On Kubernetes](http://spark.apache.org/docs/latest/running-on-kubernetes.html)
+* An active Kubernetes Cluster
+* KubeCtl
```
Comment (pan3793) [inline docs/deployment/on_kubernetes.md:15]: Kubectl

**Review by pan3793 — APPROVED** [review]
LGTM, thanks!

File: docs/deployment/on_kubernetes.md [hunk 6]
```diff
@@ -5,3 +5,111 @@
 </div>
 
 # Running Kyuubi on Kubernetes
```
Comment (yaooqinn) [inline docs/deployment/on_kubernetes.md:7]: We can rename this to `# Deploy Kyuubi engines on Kubernetes` to be more specific. we can change to yarn page too in this pr

File: docs/deployment/on_kubernetes.md [hunk 7]
```diff
@@ -5,3 +5,111 @@
...
+
+## Requirements
+
+When you want to run Kyuubi's Spark SQL engines on Kubernetes, you'd better have cognition upon the following things.
+
+* Read about [Running Spark On Kubernetes](http://spark.apache.org/docs/latest/running-on-kubernetes.html)
+* An active Kubernetes Cluster
```
Comment (yaooqinn) [inline docs/deployment/on_kubernetes.md:14]: cluster to be consistent with L16

File: docs/deployment/on_kubernetes.md [hunk 8]
```diff
@@ -5,3 +5,111 @@
...
+* Kubectl
```
Comment (yaooqinn) [inline docs/deployment/on_kubernetes.md:15]: add a ref for it to the k8s doc?

File: docs/deployment/on_kubernetes.md [hunk 9]
```diff
@@ -5,3 +5,111 @@
...
+
+## Configurations
+
+### Test Cluster
+
+You can use the shell code to test your cluster whether it is normal or not.
+
+```shell
+$SPARK_HOME/bin/spark-submit \
+ --master K8s://https://<K8s-apiserver-host>:<K8s-apiserver-port> \
```
Comment (yaooqinn) [inline docs/deployment/on_kubernetes.md:26]: K8s or k8s or both ok?

Comment (zwangsheng) [reply docs/deployment/on_kubernetes.md:26]: in this cmd, i think maybe `k8s` better

File: docs/deployment/on_kubernetes.md [hunk 10]
```diff
@@ -5,3 +5,111 @@
...
+
+`K8s://https://<K8s-apiserver-host>:<K8s-apiserver-port>`
+
+You can use cmd `kubectl cluster-info` to get api-server host and port.
+
+### Spark Image
+
+Kubernetes requires users to supply images that can be deployed into containers within pods.
+
+Spark also ships within a `./bin/docker-image-tool.sh` script that can be used to build and publish the Docker images to use
```
Comment (yaooqinn) [inline docs/deployment/on_kubernetes.md]: do we need to build the image and get the cluster-info first before telling users to run the example code?

Comment (zwangsheng) [reply docs/deployment/on_kubernetes.md]: Will adjust Master and image before test

File: docs/deployment/on_kubernetes.md [hunk 11]
```diff
@@ -4,4 +4,112 @@
...
+
+Note: Please
+see [the Security section of this document](http://spark.apache.org/docs/latest/running-on-kubernetes.html#security) for
+security issues related to volume mounts.
+
+```shell
+--conf spark.kubernetes.driver.volumes.[VolumeType].[VolumeName].mount.path=<mount path>
+--conf spark.kubernetes.driver.volumes.[VolumeType].[VolumeName].mount.readOnly=<true|false>
+--conf spark.kubernetes.driver.volumes.[VolumeType].[VolumeName].mount.subPath=<mount subPath>
+```
```
Comment (yaooqinn) [inline docs/deployment/on_kubernetes.md:103]: this (--conf) is still for the SparkPi example? Kyuubi has not supported such type for arg so far

File: docs/deployment/on_kubernetes.md [hunk 12]
```diff
@@ -4,4 +4,112 @@
...
+
+### Master
+
+Spark on Kubernetes config master by using a special format.
+
+`k8s://https://<k8s-apiserver-host>:<k8s-apiserver-port>`
+
+You can use cmd `kubectl cluster-info` to get api-server host and port.
+
+### Spark Image
```
Comment (yaooqinn) [inline docs/deployment/on_kubernetes.md:28]: docker image

File: docs/deployment/on_kubernetes.md [hunk 13]
```diff
@@ -4,4 +4,112 @@
...
+
+Kubernetes requires users to supply images that can be deployed into containers within pods.
+
+Spark also ships within a `./bin/docker-image-tool.sh` script that can be used to build and publish the Docker images to use
+with the K8s backend.
```
Comment (yaooqinn) [inline docs/deployment/on_kubernetes.md]:
```suggestion
Kubernetes requires users to supply images that can be deployed into containers within pods.

```suggestion
Spark ships a `./bin/docker-image-tool.sh` script to build and publish the Docker images for running Spark applications on Kubernetes.

When deploying Kyuubi engines against a Kubernetes cluster, we need to set up the docker images in the Docker registry first.
```

File: docs/deployment/on_kubernetes.md [hunk 14]
```diff
@@ -4,4 +4,112 @@
...
+./bin/docker-image-tool.sh -r <repo> -t my-tag push
+# To build docker image with specify openJdk 
+./bin/docker-image-tool.sh -r <repo> -t my-tag -b java_image_tag=<openjdk:${java_image_tag}> build
+# To build additional PySpark docker image
+./bin/docker-image-tool.sh -r <repo> -t my-tag -p ./kubernetes/dockerfiles/spark/bindings/python/Dockerfile build
+# To build additional SparkR docker image
+./bin/docker-image-tool.sh -r <repo> -t my-tag -R ./kubernetes/dockerfiles/spark/bindings/R/Dockerfile build
+```
+
+Of course, you can build self-made docker image by modifying Dockerfile and source code.
```
Comment (yaooqinn) [inline docs/deployment/on_kubernetes.md:48]: this is unnecessary

File: docs/deployment/on_kubernetes.md [hunk 15]
```diff
@@ -4,4 +4,118 @@
...
+* [Kubectl](https://kubernetes.io/docs/reference/kubectl/overview/)
+* KubeConfig of the target cluster
+
+## Configurations
+
+### Master
+
+Spark on Kubernetes config master by using a special format.
+
+`k8s://https://<k8s-apiserver-host>:<k8s-apiserver-port>`
```
Comment (yaooqinn) [inline docs/deployment/on_kubernetes.md:24]: spark.master=...

Comment (yaooqinn) [reply docs/deployment/on_kubernetes.md:24]: maybe we should also have a section for deply-mode

File: docs/deployment/on_kubernetes.md [hunk 16]
```diff
@@ -4,4 +4,118 @@
...
+ --conf spark.kubernetes.container.image=<spark-image> \
+ local://<path_to_examples.jar>
+```
+
+When running shell, you can use cmd `kubectl describe pod <podName>` to check if the information meets expectations.
+
+### ServiceAccount
+
+When use Client mode to submit application, spark driver use the kubeconfig to access api-service to create and watch
+executor pods.
```
Comment (yaooqinn) [inline docs/deployment/on_kubernetes.md:70]: inline

File: docs/deployment/on_kubernetes.md [hunk 17]
```diff
@@ -4,4 +4,118 @@
...
+
+When use Cluster mode to submit application, spark driver pod use serviceAccount to access api-service to create and
+watch executor pods.
```
Comment (yaooqinn) [inline docs/deployment/on_kubernetes.md:73]: inline

File: docs/deployment/on_kubernetes.md [hunk 18]
```diff
@@ -4,4 +4,118 @@
...
+
+In both cases, you need to figure out whether you have the permissions under the corresponding namespace. You can use
+following cmd to create serviceAccount (You need to have the kubeconfig which have the create serviceAccount permission)
```
Comment (yaooqinn) [inline docs/deployment/on_kubernetes.md:76]: inline

File: docs/deployment/on_kubernetes.md [hunk 19]
```diff
@@ -4,4 +4,118 @@
...
+* hostPath: mounts a file or directory from the host node’s filesystem into a pod.
+* emptyDir: an initially empty volume created when a pod is assigned to a node.
+* nfs: mounts an existing NFS(Network File System) into a pod.
+* persistentVolumeClaim: mounts a PersistentVolume into a pod.
+
+Note: Please
+see [the Security section of this document](http://spark.apache.org/docs/latest/running-on-kubernetes.html#security) for
+security issues related to volume mounts.
+
+```shell
```
Comment (yaooqinn) [inline docs/deployment/on_kubernetes.md:99]: rm shell

File: docs/deployment/on_kubernetes.md [hunk 20]
```diff
@@ -4,4 +4,118 @@
...
+Kubernetes allows defining pods from template files. Spark users can similarly use template files to define the driver
+or executor pod configurations that Spark configurations do not support.
+
+To do so, specify the spark properties `spark.kubernetes.driver.podTemplateFile` and
+`spark.kubernetes.executor.podTemplateFile` to point to local files accessible to the spark-submit process.
+
+### Other
+
+You can read [spark](http://spark.apache.org/docs/latest/running-on-kubernetes.html) for more information about spark
+configurations.
```
Comment (yaooqinn) [inline docs/deployment/on_kubernetes.md:121]:
```suggestion
You can read Spark's offical documentation for [Running on Kubernetes](http://spark.apache.org/docs/latest/running-on-kubernetes.html) for more information.
```

**Review by yaooqinn — APPROVED** [review]

**Comment by yaooqinn** [comment]
thanks, merged to master

**Pull request closed without merging**